### DISCUSSION TITLE
[Question] How many controllers do I need to run? And what happens during a migration?

### DISCUSSION THREAD

**Discussion by JohnKiller** [ceph/ceph-csi#3495] [Q&A] [answered]
Hi, I didn't find this answer on the documentation, sorry if it's already been answered somewhere.

I'm using Nomad as the orchestrator, and I would like to know how many controllers do I need to run to maintain HA.

If only 1 is needed, what happens when the orchestrator moves the controller to another host? Any downtime?

Thanks

**Comment by Madhu-1** [comment, marked as answer]
Am not exactly sure how Nomad works with storage, but CephCSI alone cannot be run in HA in the CO. it needs to be run with some sidecar, for example, external-provisioner, external-snapshotter/external-resizer which are the sidecars in the Kubernetes which works with the leader election, in Kubernetes, you can have active/passive leaders. if one leader pod goes down another leader will take over.

**Reply by JohnKiller** [reply]
In case of a node/host failure, Nomad will reschedule the controller on another node, but it will be started from scratch. Does the controller need to persist its state somewhere? Do I need to mount a volume to it and migrate that too?

**Reply by Madhu-1** [reply]
>In case of a node/host failure, Nomad will reschedule the controller on another node, but it will be started from scratch.

This is expected and fine as cephcsi is stateless.

>Does the controller need to persist its state somewhere?

I dont think it's required. It will serve the request as before and recover if something is broken during the failure.

>Do I need to mount a volume to it and migrate that too?

What do you mean by this

**Reply by JohnKiller** [reply]
Thanks for your answer. Since you said that it's stateless, the question I had about the volume it's not relevant anymore. I'll do some testing in my lab.

Thanks again

**Reply by Madhu-1** [reply]
Feel free to open issues if you find any.
